### Part 1 and 2
#### Task 1: Loading Data and Initial Inspection

In [1]:
# Import libraries
import pandas as pd
import numpy as np

In [2]:
#Load datasets
URL= "https://raw.githubusercontent.com/YBI-Foundation/Dataset/main/Bank%20Churn%20Modelling.csv"
Bank_churn_modelling="Bank_Churn_Modelling_extra_UPDATED_v2.csv"

# Put into a dataframe
churn_df= pd.read_csv( URL, sep=',')  # churn_df - bank churn dataset
model_df= pd.read_csv(Bank_churn_modelling, sep=',')  # model_df - bank churn modelling dataset csv file


In [3]:
# Compare Structures of the datasets
## churn data set
print (f"Column Names for churn dataset:", churn_df.columns)
print (churn_df.info())

## modelling data set
print (f"Column Names for modelling dataset:", model_df.columns)
print ( model_df.info())

Column Names for churn dataset: Index(['CustomerId', 'Surname', 'CreditScore', 'Geography', 'Gender', 'Age',
       'Tenure', 'Balance', 'Num Of Products', 'Has Credit Card',
       'Is Active Member', 'Estimated Salary', 'Churn'],
      dtype='str')
<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   CustomerId        10000 non-null  int64  
 1   Surname           10000 non-null  str    
 2   CreditScore       10000 non-null  int64  
 3   Geography         10000 non-null  str    
 4   Gender            10000 non-null  str    
 5   Age               10000 non-null  int64  
 6   Tenure            10000 non-null  int64  
 7   Balance           10000 non-null  float64
 8   Num Of Products   10000 non-null  int64  
 9   Has Credit Card   10000 non-null  int64  
 10  Is Active Member  10000 non-null  int64  
 11  Estimated Salary  10000 non-null  fl

#### Answers to Task 1

Two data sets have been provided, a Bank Churn data set from a URL link and a Bank Churn Modelling data set from a CSV file. The Bank Churn data set (churn set) comprises of 1000 entries per attribute and consists of 13 attributes. Meanwhile, the Bank Churn modelling data set (model set) comprises of 49 data entries per attribute with 13 attribute categories. In terms of data integrity, the churn set consists of no null-values compared to the model set which includes null-values. Therefore, the churn set is more complete and has higher integrity. 

Next, both sets are structured secondary data types due to their CSV formats, having attributes, and consisting of columns or rows, which provide a schema. Also, the data sets were acquired from a source and not directly created for this project, classifying it as secondary data as well. Both datasets also share the following attributes and data types.

|Attribute|Pandas data type|Data type by generation|Quantitative/Qualitative|Scale|
|:--------:|:--------:|:--------:|:--------:|:--------:|
|CustomerId |int64|Captured|Qualitative|Nominal|
|Surname |str|Captured|Qualitative|Nominal|
|Geography |str|Captured|Qualitative|Nominal|
|Gender |str|Captured|Qualitative|Nominal|
| Tenure |int64|Captured|Quantitative|Ratio|
| Balance |float64|Captured|Quantitative|Ratio|
| Num of products |int64|Capture|Quantitative|Ratio|
| Has Credit Card |int64|Captured|Qualitative|Nominal|
| Is Active Member |int64|Captured|Qualitative|Nominal|
| Estimated Salary |float64|Derived|Quantitative|Ratio|
| Churn |int64|Derived|Qualitative|Nominal|

However, the following two attributes have different data types across data sets: CreditScore and Age

|Data set| Attribute| Pandas data type| Data type by generation| Quantitative/Qualitative|Scale|
|:--------:|:--------:|:--------:|:--------:|:--------:|:----------:|
|churn set|CreditScore|int64|Derived|Quantitative|Interval|
|churn set|Age|int64|Captured|Quantitative|Ratio|
|model set|CreditScore|float64|Derived|Quantitative|Interval|
|model set|Age|float64|Captured|Quantitative|Ratio|



In [4]:
print (f"Look at churn set format\n", churn_df.head(10))
print (f"Look at model set format:\n", model_df.head(10))

Look at churn set format
    CustomerId   Surname  CreditScore Geography  Gender  Age  Tenure  \
0    15634602  Hargrave          619    France  Female   42       2   
1    15647311      Hill          608     Spain  Female   41       1   
2    15619304      Onio          502    France  Female   42       8   
3    15701354      Boni          699    France  Female   39       1   
4    15737888  Mitchell          850     Spain  Female   43       2   
5    15574012       Chu          645     Spain    Male   44       8   
6    15592531  Bartlett          822    France    Male   50       7   
7    15656148    Obinna          376   Germany  Female   29       4   
8    15792365        He          501    France    Male   44       4   
9    15592389        H?          684    France    Male   27       2   

     Balance  Num Of Products  Has Credit Card  Is Active Member  \
0       0.00                1                1                 1   
1   83807.86                1                0          

#### Answers to Task 1 continued...
##### Are the data sets in a long or wide format?

Both sets have the same number of columns, the same attributes but different row numbers. Both data sets may seem as long formats due to the many rows and having identifier columns but these data sets are actually in a wide format. Wide format data stores one row per unique object/entity and each variable in a separate column. Thus, both data set's are wide formats because each varible only has one row of data and multiple measurements per variable are absent.

##### The influence of strutural format?
Wide formats are usually avoided in tidy data sets but in this case a wide format is better due to the 
However, the numerical null-values in the model data set will require replacement values of either the mean, median or mode. Otherwise, aggregation operations will fail. Finally, merging the two data sets will require tidying the data by stripping spaces and different pandas data types will need to be converted for some methods, otherwise errors like truncation can occur. Truncation of float values will decrease precision of the data and decrease data integrity. 

#### Task 2: Consolidation

First, the merge() function is used when combining related data tables based on matching columns. In this project, the two dataframes have the same attributes and can use the CustomerId column as a primary key to join together, but the age and CreditScore attributes have different pandas data types. If we join the tables using merge(), selecting how="inner", then those two attribute columns (CreditScore and Age) will be kept individually and not merged with NaN values. If we select how="outer", the same outcome as how="inner" will occur. This will not provide the merged table we want and decrease data integrity by creating more NaN values and duplicates. If we want to use merge(), then the CreditScore and Age attributes will need to be converted to the same data type using astype(). However, even though this conversion could improve data integrity, there is a risk of truncation occuring with the float values becoming integers, this decreases precision. Also, the missing values (NaN's) in the model data set could fail with converting.

Next, the combine_first() function is another way to combine two data sets. It is used to combine data with the same set of entities but one data set may be missing data, this function fills in those gaps. However, for this project the goal is to combine the data sets into one long dataframe and not replace missing values with another source. This method will decrease the data integrity by filling any null-values with other customer data instead of that customers data. Hence, the accuracy of the data will be lowered and that is why this method will not be used. 

The last option would be to use concat(). Concat() is ideal for stacking data vertically, especially if is more records of the same structure. Due to the matching column records but different row numbers, stacking would work well. If we stack the churn data set (10000 rows) on top of the model data set (49 rows), we should get 10049 rows. It is similar to merge() due to also using joins but its advantage is that it aligns rows by index. The index can also be changed and the new data set can have observations (rows) with different variables(columns) because it automatically converts different data types into a compatible one for the whole column. Hence, to keep data intergity as much as possible, concat() will be used for this project.

In [5]:
# Consolidate data into one dataframe
# # Put into a dataframe
# churn_df= pd.read_csv( URL, sep=',')  # churn_df - bank churn dataset
# model_df= pd.read_csv(Bank_churn_modelling, sep=',')  # model_df - bank churn modelling dataset csv file
# clean dataframes
#df_churn= df_churn.strp()
print(f"Churn data shape:",churn_df.shape)
print(f"Model data shape:", model_df.shape)
df_combined= pd.concat([churn_df, model_df], axis=0, ignore_index=True, join='inner')
print(f"New Combined Data Set Shape:", df_combined.shape)


Churn data shape: (10000, 13)
Model data shape: (49, 13)
New Combined Data Set Shape: (10049, 13)


#### Answers to Task 2 continued...
The new combined data set now has 10049 rows, which means there should be no duplicates due to merging, but a duplicate analysis will reveal any duplicates.

#### Task 3: Duplicate Analysis and Resolution

In [6]:
#Stripping data
text_cols = ["Surname", "Geography", "Gender"]
for col in text_cols:
    df_combined[col] = df_combined[col].astype("string").str.strip().str.title()

# Checking for duplicate data
print(f"Duplicate check whole dataset:\n", df_combined.duplicated().sum())
print(f"Duplicate check CustomerId:\n", df_combined.duplicated(subset=['CustomerId']).sum())
print(f"Duplicate check for the combination of CustomerId and Credit Score:\n", df_combined.duplicated(subset=['CustomerId', 'CreditScore']).sum())
print(f"Duplicate check for a combination of CustomerId and Age:\n", df_combined.duplicated(subset=['CustomerId','Age']).sum())
print(f"Duplicate check for a combination of CustomerId, Age and Surname:\n", df_combined.duplicated(subset=['CustomerId','Age', 'Surname']).sum())
print(f"Duplicate check for a combination of CustomerId, Age, Surname and membership:\n", df_combined.duplicated(subset=['CustomerId','Age', 'Surname','Is Active Member']).sum())


Duplicate check whole dataset:
 2
Duplicate check CustomerId:
 11
Duplicate check for the combination of CustomerId and Credit Score:
 3
Duplicate check for a combination of CustomerId and Age:
 3
Duplicate check for a combination of CustomerId, Age and Surname:
 2
Duplicate check for a combination of CustomerId, Age, Surname and membership:
 2


If the whole data set is obseved for duplicates using the duplicated function, then finding duplicates is hard due to the dataset size. Therefore, the sum() function is used to count the duplicates. This method shows 2 exact duplicates across the entire dataset. However, if only the CustomerId column was used to check for duplicates because each customer is supposed to only have one ID, then the duplicated function shows that there are 11 duplicates present. When checking for duplicates with matching customer id's and age, there are 3 partial duplicates. Finally, customers with matching customer id's, ages and surnames were checked and only 2 duplicates were found. Hence, only the duplicates with an exact match will be altereted by keeping only the first record, since the record kept does not matter in this case. Then the duplicates will be re-examined. 

In [7]:
# Removing the exact duplicates only
df_dup= df_combined.drop_duplicates(subset=None, keep='first', inplace= False)
print( f"New dataframe duplicate count:", df_dup.duplicated().sum())
print(f"New dataframe shape:", df_dup.shape)

New dataframe duplicate count: 0
New dataframe shape: (10047, 13)


The duplicates were removed and the new dataframe has 10047 rows, therefore, only two duplicates were removed. Duplicates in account number can be checked again for partial duplicates. 

In [8]:
# Checking for partial duplicates
print(f"Duplicate check CustomerId:\n", df_dup.duplicated(subset=['CustomerId']).sum())
print(f"Duplicate check for the combination of CustomerId, Credit Score and Age:\n", df_dup.duplicated(subset=['CustomerId', 'CreditScore', 'Age']).sum())
print(f"Duplicate check for a combination of CustomerId and Age:\n", df_dup.duplicated(subset=['CustomerId','Age']).sum())
print(f"Duplicate check for a combination of CustomerId, Age and Surname:\n", df_dup.duplicated(subset=['CustomerId','Age', 'Surname']).sum())
print(f"Duplicate check for a combination of CustomerId, Age and Geography:\n", df_dup.duplicated(subset=['CustomerId','Age', 'Geography']).sum())
print(f"Duplicate check for a combination of CustomerId, Age, Geography and Gender:\n", df_dup.duplicated(subset=['CustomerId','Age', 'Geography', 'Gender']).sum())
print(f"Duplicate check for a combination of CustomerId, Age, Geography, Gender and Tenure:\n", df_dup.duplicated(subset=['CustomerId','Age', 'Geography', 'Gender', 'Tenure']).sum())


Duplicate check CustomerId:
 9
Duplicate check for the combination of CustomerId, Credit Score and Age:
 0
Duplicate check for a combination of CustomerId and Age:
 1
Duplicate check for a combination of CustomerId, Age and Surname:
 0
Duplicate check for a combination of CustomerId, Age and Geography:
 1
Duplicate check for a combination of CustomerId, Age, Geography and Gender:
 1
Duplicate check for a combination of CustomerId, Age, Geography, Gender and Tenure:
 0


The results from the second duplicate check only shows partial matches and even the partial duplicate for customer Id and age can be ignored because there is no partial match when more conditions like geography, gender, and tenure are added sequentially. If it was a duplicate of a customer changing their surname then the tenure and age would have matched, hence it is a partial duplicate. The only problematic duplicates remaining are the customer id's, this can be solved by changing the duplicates by adding an additional number. Microsoft Copilot provided three code options as solutions and I chose the option which will alter the data the least to keep its integrity. (https://copilot.microsoft.com/shares/ijJvosm8nrzBUoiY4JgSh)

In [9]:
#Changing duplicate Id's
df_dup['New Customer Id'] = df_dup['CustomerId'].astype(str) + '_' + df_dup.groupby('CustomerId').cumcount().astype(str)
print (df_dup.head(10))


   CustomerId   Surname  CreditScore Geography  Gender   Age  Tenure  \
0    15634602  Hargrave        619.0    France  Female  42.0       2   
1    15647311      Hill        608.0     Spain  Female  41.0       1   
2    15619304      Onio        502.0    France  Female  42.0       8   
3    15701354      Boni        699.0    France  Female  39.0       1   
4    15737888  Mitchell        850.0     Spain  Female  43.0       2   
5    15574012       Chu        645.0     Spain    Male  44.0       8   
6    15592531  Bartlett        822.0    France    Male  50.0       7   
7    15656148    Obinna        376.0   Germany  Female  29.0       4   
8    15792365        He        501.0    France    Male  44.0       4   
9    15592389        H?        684.0    France    Male  27.0       2   

     Balance  Num Of Products  Has Credit Card  Is Active Member  \
0       0.00                1                1                 1   
1   83807.86                1                0                 1   
2  

In [10]:
#### Final Duplicate check
print (f"Final Duplicate check:", df_dup.duplicated().sum())
print (f"DataFrame Size after duplicate removal:", df_dup.shape)

Final Duplicate check: 0
DataFrame Size after duplicate removal: (10047, 14)


After removal of exact duplicates and altering the partial CustomerId duplicates, the dataframe reduced by only 2 rows. The dataframe's integrity was kept due to creating a new unique key (New Customer Id) for the CustomerId duplicates, by keeping the original Id's there is tracibility for future alterations or mistakes. 

#### Task 4: Finding and Filling Missing Values


In [11]:
# First confirming the presence of missing values in each variable and the quantity
print (f"Number of missing values:\n", df_dup.isna().sum())



Number of missing values:
 CustomerId          0
Surname             0
CreditScore         2
Geography           2
Gender              1
Age                 3
Tenure              0
Balance             3
Num Of Products     0
Has Credit Card     0
Is Active Member    0
Estimated Salary    1
Churn               0
New Customer Id     0
dtype: int64


Before handling the missing data, it is important to explore the type of data missing to determine how it will be fixed. 
|Data column|Data type|Level of Measurement|Affected by|Handling options|
|:--:|:--:|:--:|:--:|:--:|
|CreditScore|Quantitative|Interval|Outliers|Mode or mean|
|Geography|Qualitative|Nominal|Nothing|Mode or Unknown|
|Gender|Qualitative|Nominal|Nothing|Mode or Unknown|
|Age|Quantitative|Ratio|Outliers|Mean or Mode|
|Balance|Quantitative|Ratio|Outliers|Mean or Mode|
|Estimated Salary|Quantitative|Ratio|Outliers|Mean or Mode|

Most of the missing data is in quantitative and will undergo statistical analysis later, therefore, any NaN values in these column indexes will be problematic. NaN values are treated differently in NumPY compared to panda's. If a NaN value is used in a statistical calculation such as the mean, then the value remains a NaN and does not get treated as a zero. Also, any NaN values need to be dealt with to avoid inaccurate or biased results. Quantitative NaN values can be replaced by either the median or mean depending on if the data is symmetrical or skewed, therefore, a slight look into the current data using the describe() function will help.

In [12]:
print(df_dup.describe())

         CustomerId   CreditScore           Age        Tenure        Balance  \
count  1.004700e+04  10045.000000  10044.000000  10047.000000   10044.000000   
mean   1.569058e+07    650.196018     38.983771      5.008062   76700.841274   
std    7.198733e+04     96.929570     10.567594      2.892533   62578.801286   
min    1.556570e+07    350.000000     18.000000      0.000000       0.000000   
25%    1.562813e+07    583.000000     32.000000      2.000000       0.000000   
50%    1.569019e+07    652.000000     37.000000      5.000000   97262.675000   
75%    1.575293e+07    717.000000     44.000000      7.000000  127730.867500   
max    1.581569e+07    850.000000     92.000000     10.000000  254145.310000   

       Num Of Products  Has Credit Card  Is Active Member  Estimated Salary  \
count     10047.000000     10047.000000      10047.000000      10046.000000   
mean          1.535881         0.704190          0.515278     100076.519275   
std           0.592341         0.456429   

According to the results, most of the columns with missing data is skewed to the left and one is skewed to the right. Therefore, the median should be used for these columns to replace the NaN values.

|Data column|Mean=Median?|Mean>Median|Mean<Median|Skewed|
|:--:|:--:|:--:|:--:|:--:|
|CreditScore|No|No|Yes|Left|
|Age|No|Yes|No|Right|
|Balance|No|No|Yes|Left|
|Estimated Salary|No|No|Yes|Left|

The NaN values in the qualitative categories geography and gender will be replaced by a new label - "unknown". This is to keep the data integrity and avoid biased results. Even though the values missing is scarce compared to the dataframe size, it is better to handle the missing values conservatively to avoid the pitfalls mentioned.

In [13]:
#Replacing NaN values 
def clean_missing_values(df_dup):
    num_cols = ['CreditScore', 'Age', 'Balance', 'Estimated Salary']
    cat_cols = ['Geography', 'Gender']
    
    
    df_dup[num_cols] = df_dup[num_cols].fillna(df_dup[num_cols].median())
    df_dup[cat_cols] = df_dup[cat_cols].fillna("Unknown")
    
    return df_dup
#Calling function
df = clean_missing_values(df_dup)



Used Copilot to help with code for replacing NaN values:
https://copilot.microsoft.com/shares/sSFwNbWqiqoQZoU22mwAf

In [14]:
# Checking NaN values available in the dataframe again
print (f"Number of missing values in cleaned data set:\n", df.isna().sum())

# Checking if dataframe still has the same shape
print (f"Cleaned DataFrame Shape:", df.shape)

Number of missing values in cleaned data set:
 CustomerId          0
Surname             0
CreditScore         0
Geography           0
Gender              0
Age                 0
Tenure              0
Balance             0
Num Of Products     0
Has Credit Card     0
Is Active Member    0
Estimated Salary    0
Churn               0
New Customer Id     0
dtype: int64
Cleaned DataFrame Shape: (10047, 14)


The new dataframe (df) still has the same amount of rows and columns as previously and there are no longer NaN values present. 

#### Task 5: Data Type and Structural Refinement
The new dataframe's datatypes will be checked again.

In [15]:
# check new dataframe datatype
print (df.dtypes)

CustomerId            int64
Surname              string
CreditScore         float64
Geography            string
Gender               string
Age                 float64
Tenure                int64
Balance             float64
Num Of Products       int64
Has Credit Card       int64
Is Active Member      int64
Estimated Salary    float64
Churn                 int64
New Customer Id         str
dtype: object


The data types for most of the column categories is correct. However, Age could be converted to an int64 to make future calculations easier because it is not a string or birthdate, therefore keeping it an interger is the best option. Also, the New Customer Id column will remain as a string due to being an identifier and to avoid any numerical operations. Therefore, the CustomerId column will be converted to the same data type as the New Customer Id column for the same reasons. Also, the gender column will be converted to categorical due to the limited options of Male, Female and Unknown, and this will reduce memory usage.

In [16]:
# Convert CustomerId column
df['CustomerId']=df['CustomerId'].astype('str')
print (df['CustomerId'].dtype)

#Convert Gender to category
df['Gender']=df['Gender'].astype('category')
print(df['Gender'].dtype)

#Convert Age to integer
df['Age']=df['Age'].astype('int')
print (df['Age'].dtype)

str
category
int64


The other categorical data categories (Has Credit Card and Is Active Member) will stay integers for possible machine learning models and will not be converted. Finally, the only column which may be removed is the surname column. The Surname column could raise privacy concerns in the project when performing analysis or machine learning algorithms and does not provide any valuable information for analysis of bank churn data. The CustomerId and New Customer Id columns could also be removed due to interferring with machine learning algorithms by causing overfitting but will be kept. The identifiers will be kept to update any data on these customers when changed in future or to compare with new data. 

In [17]:
# Removing the surnames
df=df.drop(columns='Surname')
print (df.head(5))

  CustomerId  CreditScore Geography  Gender  Age  Tenure    Balance  \
0   15634602        619.0    France  Female   42       2       0.00   
1   15647311        608.0     Spain  Female   41       1   83807.86   
2   15619304        502.0    France  Female   42       8  159660.80   
3   15701354        699.0    France  Female   39       1       0.00   
4   15737888        850.0     Spain  Female   43       2  125510.82   

   Num Of Products  Has Credit Card  Is Active Member  Estimated Salary  \
0                1                1                 1         101348.88   
1                1                0                 1         112542.58   
2                3                1                 0         113931.57   
3                2                0                 0          93826.63   
4                1                1                 1          79084.10   

   Churn New Customer Id  
0      1      15634602_0  
1      0      15647311_0  
2      1      15619304_0  
3      0      

In [18]:
# Option to Save the cleaned data set
# df.to_csv("Cleaned_Bank_Churn_data.csv", index=False)
# print("\nSaved cleaned file: Cleaned_Bank_Churn_data.csv")

#### Task 6: Customer Segmentation and Value-Based Structuring
Customers at banks can be divided into various age bands according to what is being analysed. In this project, bank churn is being analysed, therefore, splitting customers into age bands based on life phases is implemented to see the churn during phases with different needs. 
The first phase would targert young adults who are transitioning from dependence on their parents to becoming independent by going to university or working. These are individuals who require their own banking for groceries, rent etc. and targets ages 18-30. 
The next phase includes adults who are working full time and are starting families or businesses. These individuals might require car loans or mortgages, financial planning advice or credit cards and targets 31-40 year olds.  
In the 41-50 age band are individuals whom are mid-career. These individuals may want more personalised banking options and start looking more into retirement options or investments. 
The 51-60 age band include individuals in the pre-retirement stage. This target audience may want more retirement options or to expand their invesment portfolios.
Followed by the retirement stage, individuals between 61-70 are banded together. People in this stage may look for alternative wealth management to optimise their retirement potential and may leave for better products.
Finally, individuals 71+ are considered retired and seniors. This group does not really change products due to the risks involved and may require advice for their will's.

Using the above as a guideline, these age bands will be used. 

In [19]:
# Dividing customers in distinct age bands

# young_adults = df[(df['Age'] >= 18) & (df['Age'] <= 30)] # ages 18-30

# early_career = df[(df['Age'] >= 31) & (df['Age'] <= 40)] # ages 31-40

# mid_career = df[(df['Age'] >= 41) & (df['Age'] <= 50)] # ages 41-50

# pre_retire = df[(df['Age'] >= 51) & (df['Age'] <= 60)] # ages 51-60

# retired = df[(df['Age'] >= 61) & (df['Age'] <= 70)] # ages 61-70

# seniors = df[df['Age'] >= 71] # ages above 70

# Copilot Code Suggestion
def assign_age_band(age):
    if 18 <= age <= 30:
        return 'Young Adults (18–30)'
    elif 31 <= age <= 40:
        return 'Early Career (31–40)'
    elif 41 <= age <= 50:
        return 'Mid Career (41–50)'
    elif 51 <= age <= 60:
        return 'Pre-Retire (51–60)'
    elif 61 <= age <= 70:
        return 'Retired (61–70)'
    elif age >= 71:
        return 'Seniors (71+)'
    else:
        return 'Unknown'

df['AgeBand'] = df['Age'].apply(assign_age_band)

# print (young_adults) # testing code 

#### Task 6.2: Value base segmentation
Customer value to the bank can be categorised by looking at either the customer bank balance or estimated salary. For this project, the estimated salary will determine customer value. Customers will be categorised into a four level system based on what is considered below the mean, the lower mean, the higher mean and above the mean salary estimates for the current customer data base. The ranges for each category will be determined by the quartile ranges of the data, therefore, setting the standard on the bank customer demographic instead of restricting it to one region.

In [20]:
# Finding the statistics on estimated salary
print(f"Summary Statistics for Estimated Salary:", df['Estimated Salary'].describe())

Summary Statistics for Estimated Salary: count     10047.000000
mean     100076.528404
std       57530.058149
min          11.580000
25%       50909.945000
50%      100168.240000
75%      149401.960000
max      199992.480000
Name: Estimated Salary, dtype: float64


In [21]:
#Value-Based Segmentation
# set variables
# min_val= df['Estimated Salary'].min()
# max_val= df['Estimated Salary'].max()
# mean_val= df['Estimated Salary'].mean()

# below_mean = df[(df['Estimated Salary']>= min_val) & (df['Estimated Salary']<= 0.25 * mean_val)]

# lower_mean = df[(df['Estimated Salary']>= 0.25 * mean_val) & (df['Estimated Salary']<= 0.50 * mean_val)]

# upper_mean = df[(df['Estimated Salary']>= 0.50 * mean_val) & (df['Estimated Salary']<= 0.75 * mean_val)]

# above_mean = df[(df['Estimated Salary']>= 0.75 * mean_val) & (df['Estimated Salary']<= max_val)]

# print(f"Below Mean Salary Estimate:", below_mean['New Customer Id'].count())
# print(f"Lower Mean Salary Estimate:", lower_mean['New Customer Id'].count())
# print(f"Upper Mean Salary Estimate:", upper_mean['New Customer Id'].count())
# print(f"Above Mean Salary Estimate:", above_mean['New Customer Id'].count())

# Copilot code suggestion (https://copilot.microsoft.com/shares/LCMxctZMvfL5WPMLQyJwq)
min_val = df['Estimated Salary'].min()
max_val = df['Estimated Salary'].max()
mean_val = df['Estimated Salary'].mean()

def assign_salary_band(sal):
    if min_val <= sal <= 0.25 * mean_val:
        return 'Below Mean (0–25%)'
    elif 0.25 * mean_val < sal <= 0.50 * mean_val:
        return 'Lower Mean (25–50%)'
    elif 0.50 * mean_val < sal <= 0.75 * mean_val:
        return 'Upper Mean (50–75%)'
    elif 0.75 * mean_val < sal <= max_val:
        return 'Above Mean (75–100%)'
    else:
        return 'Unknown'

df['SalaryBand'] = df['Estimated Salary'].apply(assign_salary_band)

#### Task 7: Grouped Aggregation Analysis
The following aggregate questions were answered using code from Copilot (https://copilot.microsoft.com/shares/LCMxctZMvfL5WPMLQyJwq) :
1. Mean Estimated Salary by Age Group - To compare financial increases in different life phases.
2. Mean Credit Score by Churn Status - To see correlation between credit card debt and customer retention.
3. Churn rate by number of products and Estimated Salary - To see how churn rate is changes with numbers of products at different salary bands.
4. Mean Creditscore by Age band and Gender - To see which customers are more likely to be bad at paying back debt.
5. Mean Member Activity by Geography and Age - To see the spread of active members among age bands by region.
6. Mean Age and balance by Geography - To see the target age and average balance geographically.


In [22]:
# Mean Estimated Salary by Age Group
salary_by_age = df.groupby('AgeBand')['Estimated Salary'].mean()
print (f"Mean Estimated Salary by Age Group\n", salary_by_age)
# Mean CreditScore by Churn Status
avg_creditscore_by_churn_status = df.groupby('Churn')['CreditScore'].mean()
print(f"\nMean Credit Score by Churn Status\n 0 - No\n 1 - Yes\n", avg_creditscore_by_churn_status)
# Churn rate by Geography, Gender and Age group
churn_rate_via_GGA = df.groupby(['Geography', 'Gender', 'AgeBand'])['Churn'].mean()
print(f"\n Churn Rate by Geography, Gender and Age Band\n", churn_rate_via_GGA )
# Churn rate by number of products and Estimated Salary
churn_rate_by_products = df.groupby(['Num Of Products', 'SalaryBand'])['Churn'].mean()
print (f"\nChurn Rate by Number of Products and Estimated Salary\n", churn_rate_by_products)
# Mean Creditscore by Age band and Gender
mean_credit_by_age_gender = df.groupby(['AgeBand', 'Gender'])['CreditScore'].mean()
print (f"\nMean Credit Score by Age Band and Gender\n", mean_credit_by_age_gender)
# Mean Member Activity by Geography and Age
mean_ma = df.groupby(['Geography', 'AgeBand'])['Is Active Member'].mean()
print(f"\n Mean Member Activity by Geography and Age\n", mean_ma)
# Mean Age and balance by Geography
mean_age_and_balance = df.groupby('Geography')[['Age', 'Balance']].mean()
print(f"\nMean Age and Balance by Geography\n",mean_age_and_balance)

# Copilot code help: (https://copilot.microsoft.com/shares/LCMxctZMvfL5WPMLQyJwq)



Mean Estimated Salary by Age Group
 AgeBand
Early Career (31–40)     99397.389664
Mid Career (41–50)      101900.273418
Pre-Retire (51–60)       97182.225533
Retired (61–70)          98203.415265
Seniors (71+)           100315.039504
Young Adults (18–30)    100948.571575
Name: Estimated Salary, dtype: float64

Mean Credit Score by Churn Status
 0 - No
 1 - Yes
 Churn
0    651.649393
1    644.555879
Name: CreditScore, dtype: float64

 Churn Rate by Geography, Gender and Age Band
 Geography  Gender   AgeBand             
France     Female   Early Career (31–40)    0.127576
                    Mid Career (41–50)      0.330000
                    Pre-Retire (51–60)      0.643275
                    Retired (61–70)         0.337838
                    Seniors (71+)           0.078947
                    Young Adults (18–30)    0.060606
           Male     Early Career (31–40)    0.069552
                    Mid Career (41–50)      0.242105
                    Pre-Retire (51–60)      0.42346

#### Task 8: Pivot Table Construction


In [23]:
bank_churn_table = pd.pivot_table( df, values='Churn', index='Geography', aggfunc='mean')
bank_churn_table

,Churn
Geography,
France,0.161920
Germany,0.325976
Spain,0.167672
Unknown,0.500000


The mean churn rate for customers in Germany (3.26) is almost double compared to France(0.162) and Spain(0.168). Unknown regions have the highest mean churn rate of 0.5. Therefore, more German customers are lost on average compared to French or Spanish customers, according to the churn rate.